lib/puppet/functions/stdlib/ssl/extract_pkey.rb

```diff
@@ -0,0 +1,9 @@
+require 'openssl'
+# @summary
+#   Extract a public key from a certificate
+Puppet::Functions.create_function(:'stdlib::ssl::extract_pkey') do
+  def extract_pkey(certificate)
+    OpenSSL::X509::Certificate.new(certificate).public_key.to_pem
```

Comment:
have only tested this with RSA and ed22519 certificates and it it only worked with the RSA one.

Reply:
which puppetserver? maybe an old jruby?

Reply:
whatever bundler uses but also using my local ruby 2.7.4p191

```shell
$ bundle exec irb                                                                                             [15:37:24]
irb(main):002:0> require 'openssl'
=> true
irb(main):003:0> OpenSSL::X509::Certificate.new(File.read '/home/jbond/tmp/debmon/ed25519.pem')
=> #<OpenSSL::X509::Certificate: subject=#<OpenSSL::X509::Name O=Internet Widgits Pty Ltd,ST=Some-State,C=AU>, issuer=#<OpenSSL::X509::Name O=Internet Widgits Pty Ltd,ST=Some-State,C=AU>, serial=#<OpenSSL::BN:0x0000564c7a931800>, not_before=2022-06-03 12:48:56 UTC, not_after=2023-06-03 12:48:56 UTC>
irb(main):004:0> c = OpenSSL::X509::Certificate.new(File.read '/home/jbond/tmp/debmon/ed25519.pem')
irb(main):006:0> OpenSSL::X509::Certificate.new(File.read '/home/jbond/tmp/debmon/ed25519.pem').public_key.to_pem
Traceback (most recent call last):
       16: from /usr/share/rubygems-integration/all/gems/bundler-2.2.5/exe/bundle:49:in `block in <top (required)>'
       15: from /usr/share/rubygems-integration/all/gems/bundler-2.2.5/lib/bundler/cli.rb:24:in `start'
       14: from /usr/share/rubygems-integration/all/gems/bundler-2.2.5/lib/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
       13: from /usr/share/rubygems-integration/all/gems/bundler-2.2.5/lib/bundler/cli.rb:30:in `dispatch'
       12: from /usr/share/rubygems-integration/all/gems/bundler-2.2.5/lib/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
       11: from /usr/share/rubygems-integration/all/gems/bundler-2.2.5/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
       10: from /usr/share/rubygems-integration/all/gems/bundler-2.2.5/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
        9: from /usr/share/rubygems-integration/all/gems/bundler-2.2.5/lib/bundler/cli.rb:494:in `exec'
        8: from /usr/share/rubygems-integration/all/gems/bundler-2.2.5/lib/bundler/cli/exec.rb:28:in `run'
        7: from /usr/share/rubygems-integration/all/gems/bundler-2.2.5/lib/bundler/cli/exec.rb:63:in `kernel_load'
        6: from /usr/share/rubygems-integration/all/gems/bundler-2.2.5/lib/bundler/cli/exec.rb:63:in `load'
        5: from /usr/bin/irb:23:in `<top (required)>'
        4: from /usr/bin/irb:23:in `load'
        3: from /usr/lib/ruby/gems/2.7.0/gems/irb-1.2.6/exe/irb:11:in `<top (required)>'
        2: from (irb):5
        1: from (irb):6:in `rescue in irb_binding'
NoMethodError (undefined method `to_pem' for #<OpenSSL::PKey::PKey:0x0000564c7a625d00>)
Did you mean?  to_enum
irb(main):007:0>  OpenSSL::X509::Certificate.new(File.read '/home/jbond/tmp/debmon/ed25519.pem').public_key.class
=> OpenSSL::PKey::PKey
irb(main):008:0> OpenSSL::X509::Certificate.new(File.read '/home/jbond/tmp/debmon/x509.pem').public_key.to_pem
=> "-----BEGIN PUBLIC KEY-----\nMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAypa/mjZw9cL6uh4HuCoS\n0QOnfmtY5GOl5Gk+eolNUR1M2yacO+QO2FbOnGoO1MpV3AFUhocNM9jwZR8BIRpi\n+Spw2JnnlljXg5qGP9356PVMyImPJYevP7/IE5Yqr32ftBLvXsAPZDh/f9d0/RC8\nEiiwmwWuqjO9zwdy8YjKTei9+U8WU2KpsPB9zBr2yoIaId9NVr/k5mcPbHCSpmL3\nnBYCs3PFaE4Wv7X+/SLDX/V9+aLYX5jC4Gpw1nvEzfkiju4gUl5jy0lRmFckDame\nXJdnQ3q08/ujrwrJ6PFd+fMtjjwz2/XdeDt+hYViezg2GjO+h0KIG7xqXHMLhGb/\nJQIDAQAB\n-----END PUBLIC KEY-----\n"
```

but yes i suspect this is dependent on ruby and the version and implementation of the ssl library you have installed

Reply:
Make sure you also do some manual testing with a puppetserver (running puppet via jruby).  I've encountered bugs/inconsistencies with the jruby implementation of openssl before.